libp2p/transports/tls/certificate.c

```diff
@@ -959,4 +959,87 @@ void cert_free_key(cert_key_t key) {
   struct cert_key_s *k = (struct cert_key_s *)key;
   EVP_PKEY_free(k->pkey);
   free(k);
+}
+
+
+cert_error_t cert_signing_req(const char *cn, cert_key_t key, cert_buffer **csr_buffer) {
+  cert_error_t ret_code = CERT_SUCCESS;
+  X509_REQ *x509_req = NULL;
+  X509_NAME *name = NULL;
+  X509_EXTENSION *ext = NULL;
+  X509V3_CTX ctx;
+  STACK_OF(X509_EXTENSION) *exts = NULL;
+  unsigned char *der = NULL;
+  size_t der_len = 0;
+
+  if (!key || !(key->pkey)) {
+    ret_code = CERT_ERROR_NO_PUBKEY;
+    goto cleanup;
+  }
+  EVP_PKEY *pkey = key->pkey;
+
+  x509_req = X509_REQ_new();
+  if (!x509_req) {
+    ret_code = CERT_ERROR_X509_REQ_GEN;
+    goto cleanup;
+  }
+
+  if (!X509_REQ_set_pubkey(x509_req, pkey)) {
+    ret_code = CERT_ERROR_PUBKEY_SET;
+    goto cleanup;
+  }
+
+  // Build SAN extension
+  X509V3_set_ctx(&ctx, NULL, NULL, x509_req, NULL, 0);
+  char san_str[256];
```

Comment:
DNS spec (RFC 1035) limits domain names to 255 bytes in wire format, including label length prefixes and the terminating null byte. in textual form (like in a SAN), this means the domain name can be at most 253 characters. so for a SAN entry like "DNS:domain", the domain can be up to 253 bytes, and the full SAN string including "DNS:" and the null terminator needs a buffer of at least 258 bytes,  so it's probably a good idea to increase the buffer size to 258 and add a condition here to validate the length, even tho we might never run into this error, but at least will make the function correct:
Maybe something like this?
```c
if (strlen(cn) > 253) {   // "DNS:" + domain + null terminator = ~258 max
  ret_code = CERT_ERROR_CN_TOO_LONG;
  goto cleanup;
}
```